### PR TITLE
[toranj] add support for CLI based test-cases

### PR DESCRIPTION
**Pull request by abtink**
This commit updates `toranj` test framework to add support for writing
CLI based test-cases (in addition to existing model which uses OT NCP
build along with `wpantund`/`wpanctl`).

----

_**Background:**_

This PR adds support for adding CLI based toranj test-cases:
- It allows to (later) migrate some of the existing test-cases that cover certain behavior of OT core to use CLI (simplify the test and remove dependency on `wpantund`)
- The CLI based test are using simulation platform in real-time (allowing time speedup) which is different from `thread-cert` scripts which uses virtual-time and event-based simulation model.
   - This can be useful to validate the (real-time and non event-based) simulation platform implementation itself
   - And also in cases where the platform layer may not support event-based model (yet). 

This commit is also included as part of the SRPL PR and is used for testing of SRP replication behavior 
-  https://github.com/openthread/openthread/pull/7257

### CODE REVIEW

**Comment by abtink** [comment]
Quick note on this: All the tests seem to be passing ok on the PR, but since this PR is renaming the workflow test names from `toranj-15.4`, `toranj-trel`, ... to `toranj-ncp-15.4`, `toranj-ncp-trel`,...  (so we can add `toranj-cli` flavors as well), the CI github action is still waiting for `toranj-15.4` to run.

**Comment by jwhui** [comment]
@abtink , can you also update [tests/toranj/README.md](https://github.com/openthread/openthread/blob/main/tests/toranj/README.md)? Can you also include information on how `toranj` differs from `thread-cert` both technically and the use cases it addresses? I know you provided some information in https://github.com/openthread/openthread/pull/7258#issue-1088513753, but I think it would also be useful in the README.md. It might be useful to have a top-level and separate, more-specific README.md for the NCP and CLI flavors.

**Comment by abtink** [comment]
In the new push, added more documentation about `toranj-cli` vs `toranj-ncp` (and its difference/uses compared to `thread-cert` test framework). Also separated the `toranj/README.md` into two files `README_CLI.md` and `README_NCP.md` for the two modes.

**Review by superwhd — APPROVED** [review]
LGTM :+1:

File: tests/toranj/cli/cli.py [hunk 1]
```diff
@@ -0,0 +1,630 @@
...
+import pexpect
+import pexpect.popen_spawn
+import signal
+import inspect
+import weakref
+
+# ----------------------------------------------------------------------------------------------------------------------
+# Constants
+
+JOIN_TYPE_ROUTER = 'route'
```
[the remark this answers is not in the document]
Comment (abtink) [reply tests/toranj/cli/cli.py]: Thanks.

**Review by jwhui — APPROVED** [review]

**Pull request merged**